src/parser.rs

```diff
@@ -132,103 +132,198 @@ pub fn parse_lcov<T: Read>(
+                            );
+                            if let Some(f) = cur_functions.get_mut(&f_name) {
+                                f.executed |= executed != 0;
+                            } else {
+                                return Err(ParserError::Parse(format!(
+                                    "FN record missing for function {}",
+                                    f_name
+                                )));
+                            }
+                        }
+                        BRDA => {
+                            // BRDA:int,int,int,int or -
+                            if branch_enabled {
+                                let line_no = iter
+                                    .take_while(|&&c| c != b',')
+                                    .fold(0, |r, &x| r * 10 + u32::from(x - b'0'));
+                                if iter.peek().is_none() {
+                                    return Err(ParserError::InvalidRecord(format!(
+                                        "BRDA at line {}",
+                                        line
+                                    )));
+                                }
+                                let _block_number = iter
+                                    .take_while(|&&c| c != b',')
+                                    .fold(0, |r, &x| r * 10 + u64::from(x - b'0'));
+                                if iter.peek().is_none() {
+                                    return Err(ParserError::InvalidRecord(format!(
+                                        "BRDA at line {}",
+                                        line
+                                    )));
+                                }
+                                let branch_number = iter
+                                    .take_while(|&&c| c != b',')
+                                    .fold(0, |r, &x| r * 10 + u32::from(x - b'0'));
+                                if iter.peek().is_none() {
+                                    return Err(ParserError::InvalidRecord(format!(
+                                        "BRDA at line {}",
+                                        line
+                                    )));
+                                }
+                                let taken = iter
+                                    .take_while(|&&c| c != b'\n' && c != b'\r')
+                                    .fold(false, |r, &x| r || x != b'-');
+                                add_branch(&mut cur_branches, line_no, branch_number, taken);
+                            } else {
+                                iter.take_while(|&&c| c != b'\n').fold(0, |_, _| 0);
+                            }
+                        }
+                        _ => {
+                            iter.take_while(|&&c| c != b'\n').fold(0, |_, _| 0);
+                        }
                     }
                 }
-                _ => {}
             }
+        } else {
+            break;
```

Comment:
Nit: add a comment `// End of input`, otherwise one has to go back to the beginning to understand what this break does. Or, even better, you could use `while let`